## Exercises: Group and aggregate

In [1]:
import pandas as pd
import seaborn as sns
titanic = sns.load_dataset("titanic")

**Use group by and aggregation on the "titanic" dataset to answer the following questions:**

**Q1:** How many males and how many females were onboard titanic?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>males:</b> 577<br>
&nbsp;&nbsp;&nbsp;<b>females:</b> 314<br>
</details>

In [2]:
titanic
sexes = titanic.groupby("sex")
men = sexes.get_group("male")
women = sexes.get_group("female")
print(f"There were {len(men)} men and {len(women)} women on board.")

There were 577 men and 314 women on board.


**Q2:** How many males and how many females survived?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>males:</b> 109<br>
&nbsp;&nbsp;&nbsp;<b>females:</b> 233<br>
</details>

In [3]:
titanic
surviving_men = titanic.query("sex == 'male' and survived == 1 ")
surviving_women = titanic.query("sex == 'female' and survived == 1 ")

print(f"Men: {len(surviving_men)}, Women: {len(surviving_women)}.")

Men: 109, Women: 233.


In [4]:
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

**Q3:** How many people traveled in each class, and what was the total amount spent on tickets in each class?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>First:</b> 216 people (£18177)<br>
&nbsp;&nbsp;&nbsp;<b>Second:</b> 184 people (£3802)<br>
&nbsp;&nbsp;&nbsp;<b>Third:</b> 491 people (£6715)<br>
</details>

In [10]:
class_count = titanic['class'].value_counts()
total_fare = titanic.groupby('class')['fare'].sum()

print("Class Count:")
print(class_count)
print("\nTotal Fare for Each Class:")
print(total_fare)

Class Count:
class
Third     491
First     216
Second    184
Name: count, dtype: int64

Total Fare for Each Class:
class
First     18177.4125
Second     3801.8417
Third      6714.6951
Name: fare, dtype: float64


C:\Users\Mikae\AppData\Local\Temp\ipykernel_16268\6192737.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  total_fare = titanic.groupby('class')['fare'].sum()


**Q4:** Where did most people embark, and what town had the highest ratio of survivors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;Most people (644) embarked in Southampton.<br>
&nbsp;&nbsp;&nbsp;However,Cherbourg had the highest ratio of survivors (55.4%).
</details>

In [20]:
embarked_in = titanic['embark_town'].value_counts()
print(f"Most people, {embarked_in.iloc[0]} out of {len(titanic)}, embarked in {embarked_in.keys()[0]}.")

Most people, 644 out of 891, embarked in Southampton.


In [21]:
survivors_per_embarkation = titanic.groupby('embark_town')['survived'].sum()

survivors_per_embarkation / embarked_in

embark_town
Cherbourg      0.553571
Queenstown     0.389610
Southampton    0.336957
dtype: float64

**Q5:** How many family members did survivors travel with on average? And for non-survivors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Survivors:</b> 0.94 family members<br>
&nbsp;&nbsp;&nbsp;<b>Non-Survivors:</b> 0.88 family members<br><br>
&nbsp;&nbsp;&nbsp;<i>* Family members counted as: sibsp + parch</i>
</details>

In [22]:
siblingspouse = titanic.groupby('survived')['sibsp'].mean()
parentchild = titanic.groupby('survived')['parch'].mean()

siblingspouse + parentchild

survived
0    0.883424
1    0.938596
dtype: float64

**Use group by and aggregation on the "titanic" dataset to solve the following tasks:**

**T1:** Extract the following data from the dataset:
<!DOCTYPE html>
<html>
<body>

<table style="font-family: Arial, sans-serif; border-collapse: collapse; width: 50%; font-size: 12px;">
  <tr>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;"></th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">who</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_passengers</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_survivors</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">ratio_survivors</th>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">woman</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">271</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">205</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.756458</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">1</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">man</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">537</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">88</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.163873</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">2</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">child</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">83</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">49</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.590361</td>
  </tr>
</table>

</body>
</html>
<br>
<details>
<summary>Solution</summary>
<pre><code>
titanic.groupby("who").aggregate(
    number_of_passengers=pd.NamedAgg(column="survived", aggfunc="count"),
    number_of_survivors=pd.NamedAgg(column="survived", aggfunc="sum"),
    ratio_survivors=pd.NamedAgg(column="survived", aggfunc="mean"),
).sort_index(ascending=False).reset_index()
</code></pre>
</details>

In [45]:
grouped = titanic.groupby('who').agg(
    number_of_passengers=pd.NamedAgg(column='who', aggfunc='count'),
    number_of_survivors=pd.NamedAgg(column='survived', aggfunc='sum'),
)

# Calculate the 'ratio_survivors' column
grouped['ratio_survivors'] = grouped['number_of_survivors'] / grouped['number_of_passengers']

# Reindex the DataFrame to match the desired index (0, 1, 2)
grouped = grouped.reindex(['man', 'woman', 'child'])

# # Reset the index to have a DataFrame with index 0, 1, 2
grouped = grouped.reset_index()

# # Rename the columns as needed
grouped = grouped.rename(columns={'who': 'Category'})

grouped

Category  number_of_passengers  number_of_survivors  ratio_survivors
0      man                   537                   88         0.163873
1    woman                   271                  205         0.756458
2    child                    83                   49         0.590361

**T2:** Create a custom function *percent()* that can be used to aggregate survivor and get a percentage string:

<!DOCTYPE html>
<html>
<body>

<table style="font-family: Arial, sans-serif; border-collapse: collapse; width: 50%; font-size: 12px;">
  <tr>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;"></th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">who</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_passengers</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_survivors</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">percentage_survivors</th>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">woman</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">271</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">205</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">75.6 %</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">1</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">man</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">537</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">88</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">16.4 %</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">2</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">child</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">83</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">49</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">59.0 %</td>
  </tr>
</table>
</body>
</html>

<br>
<details>
<summary>Solution</summary>
<pre><code>
import numpy as np
def percent(col):
    return f"{np.mean(col) * 100:.1f} %"
</code></pre>
</details>


In [44]:
def percent(column):
    return f"{(column/len(titanic)):.1%}"

grouped = titanic.groupby('who').agg(
    number_of_passengers=pd.NamedAgg(column='who', aggfunc='count'),
    number_of_survivors=pd.NamedAgg(column='survived', aggfunc='sum'),
)

# Calculate the 'ratio_survivors' column
grouped['ratio_survivors'] = grouped['number_of_survivors'] / grouped['number_of_passengers']

# Calculate the 'percentage_survived' column using the 'percent' function,
# as that function is not a method in the dataframe, but extra.
grouped['percentage_survived'] = grouped['number_of_survivors'].apply(percent)

# Reindex the DataFrame to match the desired index (0, 1, 2)
grouped = grouped.reindex(['man', 'woman', 'child'])

# Reset the index to have a DataFrame with index 0, 1, 2
grouped = grouped.reset_index()

# Rename the columns as needed
grouped = grouped.rename(columns={'who': 'Category'})

grouped

Category  number_of_passengers  number_of_survivors  ratio_survivors  \
0      man                   537                   88         0.163873   
1    woman                   271                  205         0.756458   
2    child                    83                   49         0.590361   

  percentage_survived  
0                9.9%  
1               23.0%  
2                5.5%